doc/interpreter/octave-config.1
.\" --------------------------------------------------------------------
.\"
.\" Copyright (C) 2003-2026 The Octave Project Developers
.\"
.\" See the file COPYRIGHT.md in the top-level directory of this
.\" or <https://octave.org/copyright/>.
.\"
.\" This file is part of Octave.
.\"
.\" Octave is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" Octave is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with Octave; see the file COPYING.  If not, see
.\" <https://www.gnu.org/licenses/>.
.\"
.\" --------------------------------------------------------------------
.\"
.de Vb \" (V)erbatim (b)egin.  Use fixed width font and no justification
.ft CW
.nf
..
.de Ve \" (V)erbatim (e)nd.  Return to regular font and justification
.ft R
.fi
..
.\" --------------------------------------------------------------------
.TH octave-config 1 "April 2026" "GNU Octave"
.SH NAME
octave\-config \- GNU Octave component and library information retrieval
.SH SYNOPSIS
\fBoctave-config\fP [\fIoption\fP]\|.\|.\|.\&
.SH DESCRIPTION
.B octave\-config
is a tool to obtain directory information for \fB.oct\fP and \fB.m\fP files for
.MR octave 1 .
.SH OPTIONS
.TP
.B \-\-m\-site\-dir
Display the main directory for local, or site-specific, \fB.m\fP script files.
.TP
.B \-\-oct\-site\-dir
Display the main directory for local, or site-specific, \fB.oct\fP dynamic-link libraries.
.TP
.B \-p \fIVAR\fP\fR,\fB \-\-print \fIVAR\fP
Print the default value of the Octave configuration variable \fIVAR\fP.
Recognized variables are:

.RS
.Vb
    API_VERSION                 LOCALARCHLIBDIR
    ARCHLIBDIR                  LOCALFCNFILEDIR
    BINDIR                      LOCALOCTFILEDIR
    CANONICAL_HOST_TYPE         LOCALSTARTUPFILEDIR
    DATADIR                     LOCALVERARCHLIBDIR
    DATAROOTDIR                 LOCALVERFCNFILEDIR
    DEFAULT_PAGER               LOCALVEROCTFILEDIR
    EXEEXT                      MAN1DIR
    FCNFILEDIR                  MAN1EXT
    IMAGEDIR                    MANDIR
    INCLUDEDIR                  OCTAVE_EXEC_HOME
    INFODIR                     OCTAVE_HOME
    INFOFILE                    OCTDATADIR
    LIBDIR                      OCTDOCDIR
    LIBEXECDIR                  OCTFILEDIR
    LOCALAPIARCHLIBDIR          OCTFONTSDIR
    LOCALAPIFCNFILEDIR          OCTINCLUDEDIR
    LOCALAPIOCTFILEDIR          OCTLIBDIR
    LOCALAPIPKGDIR              STARTUPFILEDIR
                                VERSION
.Ve
.RE
.TP
.B \-v\fR,\fB \-\-version
Display the version number of
.MR octave 1 .
.TP
\fB\-h\fP, \fB\-\-help\fP
Print help message.
.SH AUTHOR
John W. Eaton and many others.  The list of contributors to the Octave project
may be shown from within \%\fBoctave\fP by executing
"\fB\%doc \%Acknowledgements\fP".  The list is also available online at
.UR https://\:\%docs.octave.org/\:\%latest/\:\%Acknowledgements\:\%.html
.UE .
.SH "SEE ALSO"
.MR octave 1 .


NEWS
Summary of important user-visible changes for version 7 (yyyy-mm-dd):
----------------------------------------------------------------------

### General improvements

- Many functions in Octave can be called in a command form---no
parentheses for invocation and no return argument assignment---or in a
functional form---parentheses and '=' for assignment of return values.

    **Command Form Example**

    `mkdir new_directory`

    **Function Form Example**

    `status = mkdir ("new_directory")`

    Octave now handles errors that occur in a consistent manner.  If
    called in command form and there is a failure, an error is thrown
    and a message printed.  If called in functional form, no error or
    message is printed and the failure is communicated to the programmer
    via the output status variable.

    The following list of functions have been modified.

    * `copyfile`
    * `fcntl`
    * `fileattrib`
    * `kill`
    * `link`
    * `mkfifo`
    * `movefile`
    * `rename`
    * `rmdir`
    * `symlink`
    * `unlink`

- Calling a user-defined function with too many inputs or outputs is now
an error.  The interpreter makes this check automatically.  If a
function uses varargin then the check is skipped for function inputs,
and if a function uses varargout then the check is skipped for function
outputs.  Input validation for functions typically begins with checking
that the number of inputs and outputs match expectations.  Existing code
can be simplified by removing these checks which are now done by the
interpreter.  Typically, code blocks like the following can simply be
deleted.

        ## Checking number of inputs
        if (nargin > 2)
          print_usage ();
        endif

        ## Checking number of outputs
        if (nargout > 1)
          print_usage ();
        endif

- Binary and hexadecimal constants like `0b101` and `0xDEADBEEF` now
create integers (unsigned by default) with sizes determined from the
number of digits present.  For example, `0xff` creates a `uint8` value
and `0xDEADBEEF` creates a `uint64` value.  You may also use a suffix of
the form `u8`, `u16`, `u32`, `u64`, `s8`, `s16`, `s32`, or `s64` to
explicitly specify the data type to use (`u` or `s` to indicate unsigned
or signed and the number to indicate the integer size).

    Binary constants are limited to 64 binary digits and hexadecimal
constants are limited to 16 hexadecimal digits with no automatic
rounding or conversion to floating point values.  Note that this may
cause problems in existing code.  For example, an expression like
`[0x1; 0x100; 0x10000]` will be uint8 (because of the rules of
concatenating integers of different sizes) with the larger values
truncated (because of the saturation semantics of integer values).  To
avoid these kinds of problems either: 1) declare the first integer to be
of the desired size such as `[0x1u32; 0x100; 0x10000]`, or 2) pad
constants in array expressions with leading zeros so that they use the
same number of digits for each value such as
`[0x00_00_01; 0x00_01_00; 0x01_00_00]`.

- The increment and decrement operators `++` and `--` must "hug" their
corresponding variables.  In previous versions of Octave, whitespaces
between these operators and the variable they affect were allowed.  That
is no longer the case.

- The `mldivide` function (i.e., the `\` operator) now uses an LU
decomposition to solve nearly singular full square matrices.  This is
Matlab-compatible and yields results which more nearly minimize `norm
(A*x - b)`.  Previously, Octave computed a minimum-norm solution.

- The `factor` function has been overhauled for speed.  For large
inputs > 1e14, it can be up to 10,000 times faster.

- The `isprime` function uses a new primality testing algorithm
that is up to 50,000 times faster for inputs > 1e14.

- The `betainc` function now calculates an exact output for the
important special cases where a or b are 1.

- The `whos` function now displays an additional attribute 's' when
the variable is a sparse type.

- As part of GSoC 2020, Abdallah K. Elshamy implemented the
`jsondecode` and `jsonencode` functions to read and write JSON data.

- As part of GSoC 2021, Abdallah K. Elshamy implemented the
`jupyter_notebook` classdef class.  This class supports running and
filling Jupyter Notebooks using the Octave language kernel from Octave
itself.  Making the evaluation of long-running Jupyter Notebooks on a
computing server without permanent browser connection possible.

- By default, the history file is now located at $DATA/octave/history,
where $DATA is a platform dependent location for (roaming) user data
files (e.g., ${XDG_DATA_HOME} or, if that is not set, ~/.local/share on
Unix-like operating systems or %APPDATA% on Windows).

- For Octave on Windows OS, the minimum required version of the Windows
API is now 6.1 (Windows 7 or newer).

- The non-re-entrant version of the QHull library "libqhull" was
deprecated upstream.  Octave now (optionally) requires the re-entrant
version of that library "libqhull_r" instead.

- Octave's libraries are now built using symbol visibility by default.
That means that less symbols are exported from these libraries.
Configure with `--disable-lib-visibility-flags` to export all symbols
(as in previous versions).

- Octave's build system no longer appends "++" to the end of the
"magick++" library name (set with the `--with-magick=` configure flag).
The real name of the "magick++" library (including any potentially
trailing "++") needs to be set in that option now.

- The `pkg update` command now accepts options that are passed to `pkg
install` for each updated package.  Specifying @option{-local} or
@option{-global} will restrict update checks to local or global
installed packages, respectively.

### Graphical User Interface

- The graphical user interface is now available in Hungarian and
Turkish.

- In debug mode, symbol values are now shown in tooltips when hovering
variables in the editor panel.


src/octave-config.in.cc
// %NO_EDIT_WARNING%

////////////////////////////////////////////////////////////////////////
//
// Copyright (C) 2008-2026 The Octave Project Developers
//
// See the file COPYRIGHT.md in the top-level directory of this
// distribution or <https://octave.org/copyright/>.
//
// This file is part of Octave.
//
// Octave is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Octave is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Octave; see the file COPYING.  If not, see
// <https://www.gnu.org/licenses/>.
//
////////////////////////////////////////////////////////////////////////

#if defined (HAVE_CONFIG_H)
#  include "config.h"
#endif

#include <string>
#include <map>
#include <iostream>
#include <algorithm>
#include <cstdlib>

#if ! defined (OCTAVE_PREFIX)
#  define OCTAVE_PREFIX %OCTAVE_PREFIX%
#endif

#if ! defined (OCTAVE_EXEC_PREFIX)
#  define OCTAVE_EXEC_PREFIX %OCTAVE_EXEC_PREFIX%
#endif

#ifndef OCTAVE_UNUSED
#  define OCTAVE_UNUSED
#endif

#include "shared-fcns.h"

static std::map<std::string, std::string> vars;

static std::string usage_msg = "usage: octave-config [options]";

static std::string help_msg =
  "\n"
  "Options:\n"
  "\n"
  "  -h, --help            Print this message.\n"
  "\n"
  "  --m-site-dir          Print the name of the directory where Octave\n"
  "                        expects to find locally installed .m files.\n"
  "\n"
  "  --oct-site-dir        Print the name of the directory where Octave\n"
  "                        expects to find locally installed .oct files.\n"
  "\n"
  "  -p VAR, --print VAR   Print the value of the given configuration\n"
  "                        variable VAR.  Recognized variables are:\n"
  "\n"
  "                          API_VERSION            LOCALARCHLIBDIR\n"    
  "                          ARCHLIBDIR             LOCALFCNFILEDIR\n"    
  "                          BINDIR                 LOCALOCTFILEDIR\n"    
  "                          CANONICAL_HOST_TYPE    LOCALSTARTUPFILEDIR\n"
  "                          DATADIR                LOCALVERARCHLIBDIR\n" 
  "                          DATAROOTDIR            LOCALVERFCNFILEDIR\n" 
  "                          DEFAULT_PAGER          LOCALVEROCTFILEDIR\n" 
  "                          EXEEXT                 MAN1DIR\n"            
  "                          FCNFILEDIR             MAN1EXT\n"            
  "                          IMAGEDIR               MANDIR\n"             
  "                          INCLUDEDIR             OCTAVE_EXEC_HOME\n"   
  "                          INFODIR                OCTAVE_HOME\n"        
  "                          INFOFILE               OCTDATADIR\n"         
  "                          LIBDIR                 OCTDOCDIR\n"          
  "                          LIBEXECDIR             OCTFILEDIR\n"         
  "                          LOCALAPIARCHLIBDIR     OCTFONTSDIR\n"        
  "                          LOCALAPIFCNFILEDIR     OCTINCLUDEDIR\n"      
  "                          LOCALAPIOCTFILEDIR     OCTLIBDIR\n"          
  "                          LOCALAPIPKGDIR         STARTUPFILEDIR\n"     
  "                                                 VERSION\n"            
  "\n"
  "  -v, --version         Print the Octave version number.\n"
  "\n";

static void
initialize ()
{
  set_octave_home ();

  vars["OCTAVE_HOME"] = Voctave_home;
  vars["OCTAVE_EXEC_HOME"] = Voctave_exec_home;

  vars["API_VERSION"] = %OCTAVE_API_VERSION%;
  vars["CANONICAL_HOST_TYPE"] = %OCTAVE_CANONICAL_HOST_TYPE%;
  vars["DEFAULT_PAGER"] = %OCTAVE_DEFAULT_PAGER%;
  vars["EXEEXT"] = %OCTAVE_EXEEXT%;
  vars["MAN1EXT"] = %OCTAVE_MAN1EXT%;
  vars["VERSION"] = %OCTAVE_VERSION%;

  vars["ARCHLIBDIR"] = prepend_octave_exec_home (%OCTAVE_ARCHLIBDIR%);
  vars["BINDIR"] = prepend_octave_exec_home (%OCTAVE_BINDIR%);
  vars["DATADIR"] = prepend_octave_home (%OCTAVE_DATADIR%);
  vars["DATAROOTDIR"] = prepend_octave_home (%OCTAVE_DATAROOTDIR%);
  vars["FCNFILEDIR"] = prepend_octave_home (%OCTAVE_FCNFILEDIR%);
  vars["IMAGEDIR"] = prepend_octave_home (%OCTAVE_IMAGEDIR%);
  vars["INCLUDEDIR"] = prepend_octave_home (%OCTAVE_INCLUDEDIR%);
  vars["INFODIR"] = prepend_octave_home (%OCTAVE_INFODIR%);
  vars["INFOFILE"] = prepend_octave_home (%OCTAVE_INFOFILE%);
  vars["LIBDIR"] = prepend_octave_exec_home (%OCTAVE_LIBDIR%);
  vars["LIBEXECDIR"] = prepend_octave_exec_home (%OCTAVE_LIBEXECDIR%);
  vars["LOCALAPIARCHLIBDIR"] = prepend_octave_exec_home (%OCTAVE_LOCALAPIARCHLIBDIR%);
  vars["LOCALAPIFCNFILEDIR"] = prepend_octave_home (%OCTAVE_LOCALAPIFCNFILEDIR%);
  vars["LOCALAPIOCTFILEDIR"] = prepend_octave_exec_home (%OCTAVE_LOCALAPIOCTFILEDIR%);
  vars["LOCALAPIPKGDIR"] = prepend_octave_exec_home (%OCTAVE_LOCALAPIPKGDIR%);
  vars["LOCALARCHLIBDIR"] = prepend_octave_exec_home (%OCTAVE_LOCALARCHLIBDIR%);
  vars["LOCALFCNFILEDIR"] = prepend_octave_home (%OCTAVE_LOCALFCNFILEDIR%);
  vars["LOCALOCTFILEDIR"] = prepend_octave_exec_home (%OCTAVE_LOCALOCTFILEDIR%);
  vars["LOCALSTARTUPFILEDIR"] = prepend_octave_home (%OCTAVE_LOCALSTARTUPFILEDIR%);
  vars["LOCALVERARCHLIBDIR"] = prepend_octave_exec_home (%OCTAVE_LOCALVERARCHLIBDIR%);
  vars["LOCALVERFCNFILEDIR"] = prepend_octave_home (%OCTAVE_LOCALVERFCNFILEDIR%);
  vars["LOCALVEROCTFILEDIR"] = prepend_octave_exec_home (%OCTAVE_LOCALVEROCTFILEDIR%);
  vars["MAN1DIR"] = prepend_octave_home (%OCTAVE_MAN1DIR%);
  vars["MANDIR"] = prepend_octave_home (%OCTAVE_MANDIR%);
  vars["OCTDATADIR"] = prepend_octave_home (%OCTAVE_OCTDATADIR%);
  vars["OCTDOCDIR"] = prepend_octave_home (%OCTAVE_OCTDOCDIR%);
  vars["OCTFILEDIR"] = prepend_octave_exec_home (%OCTAVE_OCTFILEDIR%);
  vars["OCTFONTSDIR"] = prepend_octave_home (%OCTAVE_OCTFONTSDIR%);
  vars["OCTINCLUDEDIR"] = prepend_octave_home (%OCTAVE_OCTINCLUDEDIR%);
  vars["OCTLIBDIR"] = prepend_octave_exec_home (%OCTAVE_OCTLIBDIR%);
  vars["STARTUPFILEDIR"] = prepend_octave_home (%OCTAVE_STARTUPFILEDIR%);
}

#if defined (OCTAVE_USE_WINDOWS_API) && defined (_UNICODE)
extern "C"
int
wmain (int argc, wchar_t **wargv)
{
  char **argv = convert_wargv_to_utf8 (argc, wargv);

#else
int
main (int argc, char **argv)
{
#endif
  initialize ();

  if (argc == 1)
    {
      std::cout << usage_msg << std::endl;
      return 1;
    }

  for (int i = 1; i < argc; i++)
    {
      std::string arg (argv[i]);

      if (arg == "-h" || arg == "--help")
        {
          std::cout << usage_msg << std::endl;
          std::cout << help_msg;
          return 0;
        }
      else if (arg == "--m-site-dir")
        std::cout << vars["LOCALVERFCNFILEDIR"] << std::endl;
      else if (arg == "--oct-site-dir")
        std::cout << vars["LOCALVEROCTFILEDIR"] << std::endl;
      else if (arg == "-v" || arg == "--version")
        std::cout << vars["VERSION"] << std::endl;
      else if (arg == "-p" || arg == "--print")
        {
          if (i < argc-1)
            {
              arg = argv[++i];
              std::cout << vars[arg] << std::endl;
            }
          else
            {
              std::cerr << "octave-config: " << arg
                        << " options requires argument" << std::endl;
              return 1;
            }
        }
      else
        {
          std::cerr << "octave-config: unrecognized argument " << arg
                    << std::endl;
          return 1;
        }
    }

  return 0;
}
